File: leetcode/medium/Q17_LetterCombinationsofaPhoneNumber.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        phoneletter = {"2": ["a", "b", "c"], "3": ["d", "e", "f"], "4": ["g", "h", "i"],
                       "5": ["j", "k", "l"], "6": ["m", "n", "o"], "7": ["p", "q", "r", "s"],
                       "8": ["t", "u", "v"], "9": ["w", "x", "y", "z"]}
        answer = []
        for d in digits:
            if not answer:
                for s in phoneletter[d]:
                    answer.append(s)
            else:
                tmp = []
                for a in answer:
                    for s in phoneletter[d]:
                        tmp.append(a+s)
                answer = tmp
        return answer

File: src/medium/Q17_LetterCombinationsofaPhoneNumber.java
package medium;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class Q17_LetterCombinationsofaPhoneNumber {
    static class Solution {
        List<String> comb;
        public List<String> letterCombinations(String digits) {
            comb = new ArrayList<>();
            int length = digits.length();
            List<Integer> digitList = new ArrayList<>();
            for (int i = 0; i < length; i++) {
                digitList.add(digits.charAt(i) - '0');
            }
            if(length > 0) find("", length, 0, digitList);
            return comb;
        }
        public void find(String current, int maxLength, int curIdx,List<Integer> digitList){
            if(current.length() == maxLength){
                comb.add(current);
                return;
            }
            int removed = digitList.get(curIdx);
            int max = removed == 7 || removed == 9 ? 4 : 3;
            int off = removed > 7 ? 1 : 0;
            for (int j = 0; j < max; j++) {
                int idx = ((removed-2) * 3) + off + j;
                find(current + (char)('a' + idx), maxLength, curIdx +1, digitList);
            }
        }
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        s.letterCombinations("23");
    }
}
